changeset_test.exs

```diff
@@ -0,0 +1,516 @@
+    end
+
+    attributes do
+      attribute :id, :uuid, primary_key?: true, default: &Ecto.UUID.generate/0
+      attribute :name, :string
+    end
+
+    relationships do
+      many_to_many :posts, Ash.Test.Changeset.ChangesetTest.Post,
+        through: Ash.Test.Changeset.ChangesetTest.PostCategory,
+        destination_field_on_join_table: :post_id,
+        source_field_on_join_table: :category_id
+    end
+  end
+
+  defmodule Author do
+    @moduledoc false
+    use Ash.Resource,
+      data_layer: Ash.DataLayer.Ets,
+      authorizers: [
+        Ash.Test.Authorizer
+      ]
+
+    ets do
+      private?(true)
+    end
+
+    actions do
+      read :default
+      create :default
+    end
+
+    attributes do
+      attribute :id, :uuid, primary_key?: true, default: &Ecto.UUID.generate/0
+      attribute :name, :string
+    end
+
+    relationships do
+      has_many :posts, Ash.Test.Changeset.ChangesetTest.Post, destination_field: :author_id
+    end
+  end
+
+  defmodule PostCategory do
+    @moduledoc false
+    use Ash.Resource, data_layer: Ash.DataLayer.Ets
+
+    ets do
+      private?(true)
+    end
+
+    actions do
+      read :default
+      create :default
+    end
+
+    relationships do
+      belongs_to :post, Ash.Test.Changeset.ChangesetTest.Post, primary_key?: true
+      belongs_to :category, Ash.Test.Changeset.ChangesetTest.Category, primary_key?: true
+    end
+  end
+
+  defmodule Post do
+    @moduledoc false
+    use Ash.Resource, data_layer: Ash.DataLayer.Ets
+
+    ets do
+      private?(true)
+    end
+
+    actions do
+      read :default
+      create :default
+      update :default
+    end
+
+    attributes do
+      attribute :id, :uuid, primary_key?: true, default: &Ecto.UUID.generate/0
+      attribute :title, :string
+      attribute :contents, :string
+    end
+
+    relationships do
+      belongs_to :author, Author
+
+      many_to_many :categories, Ash.Test.Changeset.ChangesetTest.Category,
+        through: Ash.Test.Changeset.ChangesetTest.PostCategory,
+        destination_field_on_join_table: :category_id,
+        source_field_on_join_table: :post_id
+    end
+  end
+
+  defmodule Api do
+    @moduledoc false
+    use Ash.Api
+
+    resources do
+      resource Category
+      resource Author
+      resource PostCategory
+      resource Post
+    end
+  end
+
+  defmodule NonResource do
+    @moduledoc false
+    defstruct [:name]
+  end
+
+  describe "new" do
+    test "it wraps a new resource in a `create` changeset" do
+      assert %changeset{
+               action_type: :create,
+               attributes: %{},
+               data: %Category{},
+               errors: [],
+               valid?: true
+             } =
+               Category
+               |> Changeset.new()
+    end
+
+    test "it wraps a resource's record in an `update` changeset" do
+      record =
+        Category
+        |> Changeset.new(%{name: "foo"})
+        |> Api.create!()
+
+      assert %changeset{
```

Comment:
```suggestion
      assert %Changeset{
```

Reply:
my bad!
even the whole file was misplaced